[SPIRV] Implements vk::BufferPointer proposal

danbrown-amd opened the pull request:
Implements [vk::BufferPointer proposal](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0010-vk-buffer-ref.md).
Closes #6489.

Keenuts commented on tools/clang/test/CodeGenSPIRV/vk.buffer-pointer.read.hlsl:
```diff
@@ -0,0 +1,48 @@
+      float4 g_vSomeConstantB;
+};
+
+typedef vk::BufferPointer<Globals_s> Globals_p;
```
Looks like doing this crashes the compiler:
```
typedef vk::BufferPointer<Globals_p> Globals_pp;;

PushConstant_t {
    Globals_pp bda;
}

bda.Get().Get().g_testFloat4

```

danbrown-amd replied:
I fixed some code that was mishandling pointer-to-pointer situations.  Does your full example work for you now?

Keenuts replied:
Yes, still can reproduce with this snippet (linux, debug build)

```
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Globals_s {
  float4 a;
};

typedef vk::BufferPointer<Globals_s> Globals_p;
typedef vk::BufferPointer<Globals_p> Globals_pp;

[[vk::push_constant]]
Globals_pp bda;

[numthreads(1, 1, 1)]
void main() {
  float4 r = bda.Get().Get().a;
}
```

```
SpirvEmitter.cpp:10865
bufferPointer->getResutlType() -> StructType
dyn_cast<SpirvPointerType>(bufferPointer->getResultType()); yields nullptr
```

Keenuts reviewed (commented):
I need to look more, early feedback

s-perron commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -1237,12 +1241,17 @@ SpirvInstruction *SpirvEmitter::doExpr(const Expr *expr,
+    if (hlsl::IsVKBufferPointerType(expr->getType())) {
+      auto *arg = constructExpr->getArg(0);
+      result = spvBuilder.createConvertUToPtr(doExpr(arg), expr->getType());
+      result->setRValue(); // ???
```
Yes, I believe you want to make this an rvalue. The constructor should not return a reference.

s-perron reviewed (commented):
This generally looks go. Thanks for getting this done. It is seems solidly implemented. I have a few concerns about the potential for future errors. With a few asserts we could address that.

I have not looked at the tests yet, nor have I tried running anything yet. I'll try to tomorrow.

danbrown-amd commented on tools/clang/include/clang/AST/HlslTypes.h:
```diff
@@ -530,6 +538,29 @@ bool DoesTypeDefineOverloadedOperator(clang::QualType typeWithOperator,
+unsigned GetVKBufferPointerAlignment(clang::QualType type);
+#endif
+
+/// <summary>Adds a constructor declaration to the specified class
+/// record.</summary> <param name="context">ASTContext that owns
+/// declarations.</param> <param name="recordDecl">Record declaration in which
+/// to add constructor.</param> <param name="resultType">Result type for
+/// constructor.</param> <param name="paramTypes">Types for constructor
+/// parameters.</param> <param name="paramNames">Names for constructor
+/// parameters.</param> <param name="declarationName">Name for
+/// constructor.</param> <param name="isConst">Whether the constructor is a
+/// const function.</param> <returns>The method declaration for the
+/// constructor.</returns>
```
I don't know, but I did it that way for consistency with the rest of the file.  Let me know if you still want it changed.

danbrown-amd commented on tools/clang/lib/AST/HlslTypes.cpp:
```diff
@@ -906,6 +909,50 @@ bool IsHLSLRayQueryType(clang::QualType type) {
+}
+
+QualType GetVKBufferPointerBufferType(clang::QualType type) {
+  return GetHLSLResourceTemplateParamType(type);
```
I combined your suggestions.

danbrown-amd commented on tools/clang/lib/SPIRV/CapabilityVisitor.cpp:
```diff
@@ -200,6 +203,7 @@ void CapabilityVisitor::addCapabilityForType(const SpirvType *type,
   }
   // Pointer type
   else if (const auto *ptrType = dyn_cast<SpirvPointerType>(type)) {
+    sc = ptrType->getStorageClass();
```
Rewritten for clarity.

danbrown-amd commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3387,6 +3420,10 @@ class HLSLExternalSource : public ExternalSemaSource {
         break;
+      case LICOMPTYPE_VK_BUFFER_POINTER:
+        paramTypes.push_back(context.getTypeDeclType(
+            m_objectTypeDecls[AR_OBJECT_VK_BUFFER_POINTER]));
```
Right.  Fixed it and verified that at least one of the new tests takes that path.

danbrown-amd commented on tools/clang/include/clang/AST/OperationKinds.h:
```diff
@@ -335,35 +340,55 @@ static_assert(
+  BO_XorAssign,
   BO_OrAssign,
-  BO_Comma                      // [C99 6.5.17] Comma operator.
+  BO_Comma // [C99 6.5.17] Comma operator.
```
Fixed.  I had a pre-commit hook in place that helpfully formatted all changed files (i.e., not just the changes); I'd already fixed most of them but this one slipped past me.

danbrown-amd commented on tools/clang/include/clang/AST/OperationKinds.h:
```diff
@@ -335,35 +340,55 @@ static_assert(
+  UO_LNot, // [C99 6.5.3.3] Unary arithmetic
+  UO_Real,
+  UO_Imag,     // "__real expr"/"__imag expr" Extension.
+  UO_Extension // __extension__ marker.
```
Also fixed.

danbrown-amd commented on tools/clang/lib/AST/HlslTypes.cpp:
```diff
@@ -906,6 +909,50 @@ bool IsHLSLRayQueryType(clang::QualType type) {
+      }
+    }
+  }
+  return false;
```
Done.

danbrown-amd commented on tools/clang/lib/AST/HlslTypes.cpp:
```diff
@@ -906,6 +909,50 @@ bool IsHLSLRayQueryType(clang::QualType type) {
+    }
+  }
+
+  return 0; // TODO: error
```
Refactored.

danbrown-amd commented on tools/clang/lib/Lex/PPMacroExpansion.cpp:
Restored original formatting.

danbrown-amd commented on tools/clang/lib/SPIRV/AlignmentSizeCalculator.cpp:
```diff
@@ -277,14 +280,18 @@ std::pair<uint32_t, uint32_t> AlignmentSizeCalculator::getAlignmentAndSize(
+        const uint64_t alignment =
+            templateDecl->getTemplateArgs()[2].getAsIntegral().getZExtValue();
+        return {alignment, size};
+      }
```
Changed as requested.

danbrown-amd commented on tools/clang/lib/SPIRV/LowerTypeVisitor.cpp:
```diff
@@ -809,6 +812,27 @@ const SpirvType *LowerTypeVisitor::lowerVkTypeInVkNamespace(
+        spirvType, spv::StorageClass::PhysicalStorageBuffer);
+    spvContext.registerForwardReference(type, pointerType);
+
+    visited.pop_back();
```
Done and done.

danbrown-amd commented on tools/clang/lib/SPIRV/SpirvBuilder.cpp:
```diff
@@ -202,6 +205,14 @@ SpirvInstruction *SpirvBuilder::createLoad(QualType resultType,
   instruction->setLayoutRule(pointer->getLayoutRule());
   instruction->setRValue(true);
 
+  if (pointer->getStorageClass() == spv::StorageClass::PhysicalStorageBuffer) {
```
Now the loads do too.

danbrown-amd commented on tools/clang/lib/SPIRV/EmitVisitor.cpp:
```diff
@@ -2581,6 +2608,17 @@ uint32_t EmitTypeHandler::emitType(const SpirvType *type) {
   }
+  // Forward pointer types
+  else if (const auto *fwdPtrType = dyn_cast<ForwardPointerType>(type)) {
+    const SpirvPointerType *ptrType =
+        context.getForwardReference(fwdPtrType->getPointeeType());
```
Forward references are now created only for vk::BufferPointer and thus always have storage class PhysicalStorageBuffer.

danbrown-amd commented on tools/clang/lib/SPIRV/SpirvContext.cpp:
```diff
@@ -328,6 +331,26 @@ const HybridPointerType *SpirvContext::getPointerType(QualType pointee,
+}
+
+void SpirvContext::registerForwardReference(
+    QualType type, const SpirvPointerType *pointerType) {
```
Done.

danbrown-amd commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -1237,12 +1241,17 @@ SpirvInstruction *SpirvEmitter::doExpr(const Expr *expr,
+      auto *arg = constructExpr->getArg(0);
+      result = spvBuilder.createConvertUToPtr(doExpr(arg), expr->getType());
+      result->setRValue(); // ???
+      // probably actually need a special case for assignment to ptr-to-ptr var
```
I was actually thinking of something a bit different but worded the comment poorly.  Anyway, the constructor argument here will always be of integral type -- though it may have to be loaded from a variable, which I had overlooked but have now accounted for.  (I realized that I didn't have a test for this constructor, so I've added one to vk.buffer-pointer.linked-list.)

s-perron reviewed (commented):
I have a suggestion for a a few extra tests. Once they are in, I feel comfortable approving. 

Can you also add a tests where an atomic operation is done? We want to be sure that works too.

Is the `has_feature` uses anywhere? We should make sure that works.

danbrown-amd commented on tools/clang/lib/Lex/PPMacroExpansion.cpp:
```diff
@@ -1080,7 +1083,8 @@ static bool HasFeature(const Preprocessor &PP, const IdentifierInfo *II) {
       .Case("memory_sanitizer", LangOpts.Sanitize.has(SanitizerKind::Memory))
       .Case("thread_sanitizer", LangOpts.Sanitize.has(SanitizerKind::Thread))
-      .Case("dataflow_sanitizer", LangOpts.Sanitize.has(SanitizerKind::DataFlow))
+      .Case("dataflow_sanitizer",
+            LangOpts.Sanitize.has(SanitizerKind::DataFlow))
```
I would have thought so, but for some reason it's necessary to pass the check.

danbrown-amd commented on tools/clang/test/CodeGenSPIRV/vk.buffer-pointer.linked-list.hlsl:
```diff
@@ -0,0 +1,94 @@
+[numthreads(1,1,1)]
+float4 MainPs(void) : SV_Target0
+{
+      [[vk::aliased_pointer]] block_p g_p =
```
It's not clear to me what is supposed to happen when the annotation is put on a struct member.  Is a decoration supposed to be generated for the corresponding field of every declared variable or function parameter of that struct?  I'm also not sure exactly what "global variable that becomes part of a cbuffer" means.

For now, I have rectified the omission of decorations on function parameters and provided a new test file for aliases, based on Appendix E of the proposal.  I've also provided tests for the other functionality requested above.

s-perron replied:
> It's not clear to me what is supposed to happen when the annotation is put on a struct member.

That is a good call. I assumed you could use it in an OpMemberDecorate instruction, but you cannot do that. We might need to think about that and maybe update the spec. It says that the attribute can be applied to a struct member.

> I'm also not sure exactly what "global variable that becomes part of a cbuffer" means.

See https://godbolt.org/z/vPEYd74nP. Notice that in this case `a` becomes a member struct:

```
OpMemberName %type__Globals 0 "a"
```

How will the attribute be applied in this case? We also need to think about how to handle it if there is one member that is aliased, and another that is restrict. 

I don't think you have to deal with those in this PR because I think they could be problems with the spec. We can deal with this in follow up PRs.

s-perron replied:
I have opened https://github.com/microsoft/hlsl-specs/issues/433 to follow up on this.

danbrown-amd commented on tools/clang/test/CodeGenSPIRV/vk.buffer-pointer.linked-list.hlsl:
```diff
@@ -0,0 +1,94 @@
+          vk::static_pointer_cast<block_t, 16>(g_PushConstants.root);
+      g_p = g_p.Get().next;
+      uint64_t addr = (uint64_t)g_p;
+      block_p copy = block_p(addr);
```
There is no real `operator=` from the compiler's point of view, only a special case of assignment.  I have added an invocation of the copy constructor.

Keenuts reviewed (commented):
Thanks for the work!
Good improvements, but looks like there are still some failure cases:

Seems like I also reach an assert when this snippet:

```hlsl
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Content {
  float a;
};

typedef vk::BufferPointer<Content> BufferContent;
typedef vk::BufferPointer<BufferContent> BufferBufferContent;

[[vk::push_constant]]
BufferContent buffer;

[numthreads(1, 1, 1)]
void main() {
  float tmp = buffer.Get().a;
  buffer.Get().a = tmp;
}
```

Keenuts commented:
Also get an assert/validation errors with those 2 snippets:

```hlsl
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Content {
  uint a;
};

typedef vk::BufferPointer<uint> BufferContent;

[[vk::push_constant]]
BufferContent buffer;

[numthreads(1, 1, 1)]
void main() {
  uint data = buffer.Get();
  buffer.Get() = data;
}
```
```
Error: assert(IsPtr(ptr_id) && "Cannot get the variable when input is not a pointer.")
(assert in SPIRV-Tools)
```

Looks like removing optimizations or changing the snippet to this shows the validation error instead:
```hlsl
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Content {
  uint a;
};

typedef vk::BufferPointer<uint> BufferContent;

[[vk::push_constant]]
BufferContent buffer;

[numthreads(1, 1, 1)]
void main() {
  buffer.Get() = 1;
}
```

Keenuts commented:
Found another case:
```hlsl
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Content {
  int a;
};

typedef vk::BufferPointer<Content> BufferContent;
typedef vk::BufferPointer<BufferContent> BufferBuffer;

//[[vk::push_constant]]
//BufferContent buffer;

RWStructuredBuffer<BufferBuffer> rwbuf;

void foo(BufferContent bc) {
  bc.Get().a = 1;
}

[numthreads(1, 1, 1)]
void main() {
  foo(rwbuf[0].Get());
}
```

```
fatal error: generated SPIR-V is invalid: [VUID-StandaloneSpirv-PhysicalStorageBuffer64-04708] Memory accesses with PhysicalStorageBuffer must use Aligned.
  %21 = OpLoad %_ptr_PhysicalStorageBuffer_Content %20
```

Keenuts commented:
Looks like casting would also need additional restrictions:

```hlsl
// RUN: %dxc -E main -T cs_6_7  %s -spirv -E main | FileCheck %s

struct Content {
  int a;
};

typedef vk::BufferPointer<Content> BufferContent;
typedef vk::BufferPointer<BufferContent> BufferBuffer;

RWStructuredBuffer<BufferContent> buf;

void foo(const BufferContent bc) {
  bc.Get().a = 1;
}

[numthreads(1, 1, 1)]
void main() {
  static BufferContent bcs = buf[0];
  static BufferBuffer bbs = (BufferBuffer)bcs;
}
```

```
fatal error: generated SPIR-V is invalid: Expected unsigned int scalar type as Result Type: ConvertPtrToU
  %19 = OpConvertPtrToU %_ptr_PhysicalStorageBuffer__ptr_PhysicalStorageBuffer_Content %18
```

danbrown-amd commented:
Should it actually be legal to have buffer pointer push constants?  The front end rejects non-struct push constant declarations., and although values of type vk::BufferPointer<T> are objects, they don't have data members and thus aren't treated as records.

Keenuts commented:
> Should it actually be legal to have buffer pointer push constants?

Mmmm, I don't think there is something that forbids it since we allow it for ConstantBuffers (at least we have one test checking this `vk.push-constant.constantbuffer.assign.hlsl`)
(@s-perron in case I'm saying something wrong)

What frontend are you referring to?

s-perron commented:
> > Should it actually be legal to have buffer pointer push constants?
> 
> Mmmm, I don't think there is something that forbids it since we allow it for ConstantBuffers (at least we have one test checking this `vk.push-constant.constantbuffer.assign.hlsl`) (@s-perron in case I'm saying something wrong)
> 
> What frontend are you referring to?

I believe the issue is that Vulkan requires that the type of a [push constant must be a struct](https://docs.vulkan.org/spec/latest/chapters/interfaces.html#interfaces-resources-pushconst). That is enforced by the compiler: https://godbolt.org/z/8r38xss5o. Since the vk::BufferPointer is a pointer and not a struct, we should not allow it as a push constant unless it is contained in another struct.

danbrown-amd commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -10770,6 +10834,48 @@ SpirvEmitter::processIntrinsicClamp(const CallExpr *callExpr) {
+  Expr *obj = callExpr->getImplicitObjectArgument();
+  SpirvInstruction *bufferPointer = doExpr(obj);
+  unsigned align = hlsl::GetVKBufferPointerAlignment(obj->getType());
+  lowerTypeVisitor.visitInstruction(bufferPointer);
```
Changed as suggested.

Keenuts reviewed (commented):
Hello! Thanks for the changes!
I'm fine moving forward with this PR so we can find more bugs with the first users, but I'd like for the fixed bugs to have test cases.
(I provided snippets/run lines, so making them into tests is mostly adding a `2>&1 | FileCheck` and `CHECK: <expected-error>`)

danbrown-amd commented:
I've pushed the requested tests.

I've rebased on the most recent main and fixed the conflicts locally, but a recent (March 4) change to SPIRV-Tools causes non-termination on the linked list example.  It appears that `Handler_OpTypePointer_StorageBuffer16BitAccess()` in trim_capabilities_pass.cpp doesn't account for recursive types.

Keenuts commented:
> Handler_OpTypePointer_StorageBuffer16BitAccess

Ah right, thanks for finding, sent https://github.com/KhronosGroup/SPIRV-Tools/pull/6070 to fix the issue.

Keenuts commented:
Once https://github.com/microsoft/DirectXShaderCompiler/pull/7269 is merged you'll be able to rebase  👍️

amaiorano commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1232,6 +1237,9 @@ static const ArBasicKind g_AnyOutputRecordCT[] = {
 static const ArBasicKind g_DxHitObjectCT[] = {AR_OBJECT_HIT_OBJECT,
                                               AR_BASIC_UNKNOWN};
 
+static const ArBasicKind g_VKBufferPointerCT[] = {AR_OBJECT_VK_BUFFER_POINTER,
```
We're getting build breaks in Chromium where ENABLE_SPIRV_CODEGEN is not defined. I think this decl should be wrapped in `#ifdef ENABLE_SPIRV_CODEGEN`?

dneto0 replied:
I'm working on a patch.

danbrown-amd replied:
Created [pull request](https://github.com/microsoft/DirectXShaderCompiler/pull/7306).
